src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -509,25 +506,34 @@ static NativeObjectWrapper()
                 ComAwareWeakReference.InitializeCallbacks(&ComWeakRefToObject, &PossiblyComObject, &ObjectToComWeakRef);
             }
 
-            public static NativeObjectWrapper Create(IntPtr externalComObject, IntPtr inner, ComWrappers comWrappers, object comProxy, CreateObjectFlags flags)
+            public static NativeObjectWrapper Create(IntPtr externalComObject, IntPtr inner, ComWrappers comWrappers, object comProxy, GCHandle proxyHandle, CreateObjectFlags flags)
             {
+                Debug.Assert(comProxy == proxyHandle.Target);
                 if (flags.HasFlag(CreateObjectFlags.TrackerObject) &&
                     Marshal.QueryInterface(externalComObject, IID_IReferenceTracker, out IntPtr trackerObject) == HResults.S_OK)
                 {
-                    return new ReferenceTrackerNativeObjectWrapper(externalComObject, inner, comWrappers, comProxy, flags, trackerObject);
+                    return new ReferenceTrackerNativeObjectWrapper(externalComObject, inner, comWrappers, comProxy, flags, trackerObject)
+                    {
+                        ProxyHandle = proxyHandle
+                    };
                 }
                 else
                 {
-                    return new NativeObjectWrapper(externalComObject, inner, comWrappers, comProxy, flags);
+                    return new NativeObjectWrapper(externalComObject, inner, comWrappers, comProxy, flags)
+                    {
+                        ProxyHandle = proxyHandle
+                    };
                 }
             }
 
-            public NativeObjectWrapper(IntPtr externalComObject, IntPtr inner, ComWrappers comWrappers, object comProxy, CreateObjectFlags flags)
+            internal required GCHandle ProxyHandle { get; init; }
```

Comment:
+1 to Jan's comment. It feels a bit weird to have fields for everything, and then just this one standalone property somewhere else (on top of the fact this introduces a copy, so you don't mutate the field anymore). If feels just leaving it as a readonly field and having the constructor parameter would be more consistent with the other parameters?